javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll

```diff
@@ -0,0 +1,161 @@
+   * The flag `secure` is not set by default (https://github.com/expressjs/session#cookiesecure).
+   * The default value for cookie options is { path: '/', httpOnly: true, secure: false, maxAge: null }.
+   */
+  class InsecureExpressSessionCookie extends ExpressLibraries::ExpressSession::MiddlewareInstance,
+    InsecureCookies {
+    override string getKind() { result = "express-session" }
+
+    override DataFlow::SourceNode getCookieOptionsArgument() {
+      result = this.getOption("cookie").(DataFlow::SourceNode)
+    }
+
+    DataFlow::Node getCookieFlagValue(string flag) {
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
+      not exists(DataFlow::SourceNode cookieOptions |
+        cookieOptions = this.getCookieOptionsArgument() and
+        getCookieFlagValue(flag()).mayHaveBooleanValue(true)
+      ) and
+      not getCookieFlagValue(flag()).mayHaveStringValue("auto")
+    }
+  }
+
+  /**
+   * A cookie set using `response.cookie` from `express` module (https://expressjs.com/en/api.html#res.cookie).
+   */
+  class InsecureExpressCookieResponse extends InsecureCookies {
+    InsecureExpressCookieResponse() {
+      this = any(Express::ResponseExpr response).flow().getALocalSource().getAMemberCall("cookie")
+    }
+
+    override string getKind() { result = "response.cookie" }
+
+    override DataFlow::SourceNode getCookieOptionsArgument() {
+      result = this.(DataFlow::InvokeNode).getLastArgument().getALocalSource()
+    }
+
+    DataFlow::Node getCookieFlagValue(string flag) {
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
+      not exists(DataFlow::SourceNode cookieOptions |
+        cookieOptions = this.getCookieOptionsArgument() and
+        getCookieFlagValue(flag()).mayHaveBooleanValue(true)
+      )
+    }
+  }
+
+  /**
+   * A cookie set using `Set-Cookie` header of an `HTTP` response (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie).
+   */
+  class InsecureSetCookieHeader extends InsecureCookies {
+    InsecureSetCookieHeader() {
+      this.asExpr() = any(HTTP::SetCookieHeader setCookie).getHeaderArgument()
+    }
+
+    override string getKind() { result = "set-cookie header" }
+
+    override DataFlow::Node getCookieOptionsArgument() {
+      result.asExpr() = this.asExpr().(ArrayExpr).getAnElement()
+    }
+
+    // A cookie is insecure if the 'secure' flag is not specified in the cookie definition.
+    override predicate isInsecure() {
+      not exists(string s |
+        getCookieOptionsArgument().mayHaveStringValue(s) and
+        s.matches("%; secure%")
```

Comment:
Hmm. What if the string is `...;secure...`? That is: it lacks a space between `;` and `secure`.

How about something like`s.regexpMatch("(.*;)?\\s*secure.*")` instead? That should allow any whitespace variation, and also permit the string to start with `secure`.

Reply:
Again, you are right. I missed that. I made the change